src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,138 @@
+        this.size = 0;
+        this.innerArray = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
+        if (size + 1 >= innerArray.length) {
+            addCapacity();
+        }
 
+        innerArray[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexValidity(index, true);
+
+        if (size + 1 >= innerArray.length) {
+            addCapacity();
+        }
 
+        moveElementsForward(index, innerArray);
+        innerArray[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidity(index);
+
+        return innerArray[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndexValidity(index);
 
+        innerArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexValidity(index);
+
+        T elementToRemove = innerArray[index];
+        moveElementsBackwards(index, innerArray);
+
+        size--;
+
+        return elementToRemove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int elementIndex = findElementIndex(element, innerArray);
+
+        if (elementIndex == -1) {
+            throw new NoSuchElementException("Invalid element");
+        }
+
+        return remove(elementIndex);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void addCapacity() {
+        int newCapacity = (int) (innerArray.length * CAPACITY_MULTIPLIER);
+        T[] newArray = (T[]) new Object[newCapacity];
+        System.arraycopy(innerArray, 0, newArray, 0, innerArray.length);
+        innerArray = newArray;
+    }
+
+    private void moveElementsBackwards(int index, T[] innerArray) {
+        for (int i = index + 1; i < innerArray.length; i++) {
+            innerArray[i - 1] = innerArray[i];
+        }
+    }
+
+    private void moveElementsForward(int index, T[] innerArray) {
+        for (int i = innerArray.length - 1; i > index; i--) {
+            innerArray[i] = innerArray[i - 1];
+        }
+    }
+
+    private int findElementIndex(T element, T[] innerArray) {
+        for (int i = 0; i < innerArray.length; i++) {
+            if (Objects.equals(innerArray[i], element)) {
+                return i;
+            }
+        }
+
+        return ELEMENT_NOT_FOUND;
+    }
+
+    private void checkIndexValidity(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
```

Comment:
Don't complicate if-else construction. The checkIndexValidity method is duplicated with only a slight variation for the add operation. Refactor this method to avoid duplication and handle the add operation case within a single method.